### DISCUSSION TITLE
USB external keyboard webADB

### DISCUSSION THREAD

**Discussion by CorruptionVR** [yume-chan/ya-webadb#475] [Q&A] [answered]
would it be possible to register the webADB client as a USB keyboard so that the onscreen keyboard doesn't show up?

**Comment by yume-chan** [comment, marked as answer]
It's planned, using the AoA protocol. I have https://github.com/9elements/aoakvm and Scrcpy's HID input mode for reference. But have no time to actually implement it. Currently working on Scrcpy screen recording feature (mostly done).

**Reply by CorruptionVR** [reply]
oh ok, thanks

**Comment by yume-chan** [comment]
Also try https://github.com/Genymobile/scrcpy/issues/767#issuecomment-1043618316. Haven't tested it. If it works well maybe I can integrate it into Scrcpy server, so the keyboard is restored even if ADB is disconnected.

**Comment by yume-chan** [comment]
Please test the AOA injection mode on https://ya-webadb-git-feat-aoa-yume-chan.vercel.app/, there are two new settings (off by default)

![image](https://user-images.githubusercontent.com/1330321/216761357-db1e225d-00d5-4bdf-bba9-132bced8f303.png)

**Reply by CorruptionVR** [reply]
oh wait, the mouse doesn't like it if I go really fast.

**Reply by CorruptionVR** [reply]
looks like it just doesn't capture ctrl

**Reply by CorruptionVR** [reply]
Problems with AoA protocol test implementation

- meta key ignored
- scrolling is choppy on chromeOS
- with great mouse speed comes great jitter
- first click to enable pointer lock registers as click, causing possible misclicks

may list more bugs as I encounter them

suggestions

- if both new options or keyboard emulation is enabled, then when you click on the canvas to focus the keyboard, automatically enter fullscreen

**Reply by yume-chan** [reply]
Keyboard lock and AOA keyboard are not related. It can also capture and send keys like alt + tab using software injection, as long as keyboard lock is enabled.

**Reply by yume-chan** [reply]
Keyboard lock only works in fullscreen mode. If it still doesn't work for you in fullscreen, please file a new issue with a screen record. You can use online tools like https://gifcap.dev/ to capture your screen.

**Reply by CorruptionVR** [reply]
I realized my mistake and deleted the comment. the keyboard lock works fine with the exception of not being able to capture the meta key, but that keystroke may not be implemented yet in the web demo

**Reply by CorruptionVR** [reply]
hey, just used a keyboard tester, and it seems that the (') key registers as the (~) key.

**Reply by CorruptionVR** [reply]
quick update, the keyboard.lock() API is broken

**Reply by yume-chan** [reply]
As I said keyboard lock and HID keyboard are not related, so I separated it into #498.

The plan is:

1. Keyboard lock will be enabled in full screen by default, without an option.
2. The only benefits of HID keyboard mode are hiding Android IME and presumably lower input latency, but it doesn't have any clear drawback, so it will be enabled on USB connection, without an option.
3. I can't find any usage of HID mouse mode (and its mouse capture), but there are so many disadvantages. Because the cursor is drawn on device, the perceived latency depends on the video latency. Press ESC will exit mouse capture, so it will require Keyboard lock, which is only available in full screen. As a result, HID mouse mode will be removed.

**Reply by CorruptionVR** [reply]
the HID mouse mode can be used for games such as Minecraft where you have the option to use a mouse to play like on a pc.
